File: src/main/java/me/deecaad/weaponmechanicsplus/weapon/modifiers/ExplosionModifier.java
package me.deecaad.weaponmechanicsplus.weapon.modifiers;

import me.deecaad.core.file.SerializeData;
import me.deecaad.core.file.Serializer;
import me.deecaad.core.file.SerializerException;
import me.deecaad.weaponmechanics.weapon.explode.Explosion;
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.ModifierHelper;
import org.jetbrains.annotations.NotNull;

public class ExplosionModifier implements Serializer<ExplosionModifier>, ModifierHelper {

    private Explosion overrideExplosion;

    /**
     * Default constructor for serializer
     */
    public ExplosionModifier() {
    }

    public ExplosionModifier(Explosion overrideExplosion) {
        this.overrideExplosion = overrideExplosion;
    }

    @NotNull
    @Override
    public ExplosionModifier serialize(SerializeData data) throws SerializerException {
        Explosion overrideExplosion = data.of("Override_Explosion").serialize(Explosion.class);

        return new ExplosionModifier(overrideExplosion);
    }
}

File: src/main/kotlin/me/deecaad/weaponmechanicsplus/weapon/modifiers/ExplosionModifier.kt
package me.deecaad.weaponmechanicsplus.weapon.modifiers

import me.deecaad.core.file.*
import me.deecaad.core.mechanics.Mechanics
import me.deecaad.weaponmechanics.weapon.explode.Explosion
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.MechanicsModifier
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.MechanicsModifier.Companion.serializeMechanicsModifier

class ExplosionModifier : Serializer<ExplosionModifier> {

    var overrideExplosion: Explosion? = null
    var mechanicsModifier: MechanicsModifier? = null

    /**
     * Default constructor for serializer
     */
    constructor()

    constructor(overrideExplosion: Explosion?, mechanicsModifier: MechanicsModifier?) {
        this.overrideExplosion = overrideExplosion
        this.mechanicsModifier = mechanicsModifier
    }

    @Throws(SerializerException::class)
    override fun serialize(data: SerializeData): ExplosionModifier {
        val overrideExplosion = data.of("Override_Explosion").serialize(Explosion::class.java)
        val mechanicsModifier = data.serializeMechanicsModifier()

        return ExplosionModifier(overrideExplosion, mechanicsModifier)
    }
}